### PR TITLE
refactor(core) client cache separation

### PR DESCRIPTION
**Pull request by gszr**
### Summary

New cache `kong.core_cache` is meant for core usage, while the existing
`kong.cache` is used in, e.g., plugins.

### CODE REVIEW

File: kong/cache.lua [hunk 1]
```diff
@@ -112,8 +112,12 @@ local mt = { __index = _M }
 
 
 function _M.new(opts)
-  if _init then
-    error("kong.cache was already created", 2)
+  if not opts.shm_name or type(opts.shm_name) ~= "string" then
```
Comment (javierguerragiraldez) [inline kong/cache.lua]: AFAICT, this is equivalent to just `if type(opts.shm_name) ~= 'string' then...`.  also note that there's no performance advantage in checking the truthyness value first, the type-specific nature of compiled traces allow type checks to be compiled away into a simple constant (but that's only relevant in the tightest loops).

**Comment by hutchic** [comment]
Are there any tests for `kong.cache`?

File: kong/cache.lua (hunk 1 above)
Comment (dndx) [reply kong/cache.lua]: Yes. `type(opts.shm_name) ~= "string"` already implies `opts.shm_name` can not be `false` or `nil`.

Comment (gszr) [reply kong/cache.lua]: Good catch @javierguerragiraldez - thanks!

**Comment by kikito** [comment]
Hey @gszr , as Colin is saying: a lot of instances of `kong.cache` where transformed into `kong.core_cache` with this change, including specs. Since now we have two kinds of cache (core and non-core), is the non-core cache tested anywhere? Perhaps on the plugin tests themselves?

**Comment by gszr** [comment]
@kikito @hutchic Yes, plugins tests - e.g., cache invalidation tests - exercise the client-facing cache.

File: kong/cache_warmup.lua [hunk 2]
```diff
@@ -66,7 +66,7 @@ local function cache_warmup_single_entity(dao)
 
     local cache_key = dao:cache_key(entity)
 
-    local ok, err = kong.cache:safe_set(cache_key, entity)
+    local ok, err = kong.core_cache:safe_set(cache_key, entity)
```
[the remark this answers is not in the document]
Comment (hishamhm) [reply kong/cache_warmup.lua]: @gszr an easy way to fix this is to either check if the dao name is in `kong.constants.CORE_ENTITIES` or to add a `is_core_entity = true` flag in the core entity DAO objects as they are created in `kong.db.init` and then check it here, and use the appropriate cache.

[the remark this answers is not in the document]
Comment (gszr) [reply kong/cache_warmup.lua]: Fix - and rebase - incoming!

**Pull request merged**